### PR TITLE
Habits v2: adding counter to habits (cleaned up branch) - fixes #8113

### LINKED ISSUE #8113: Habits v2 - expected frequency/counter/last time performed (from Trello)
In the trello card [HABITS v2 (expected Habit Frequency, Onlies, Habit Maximum, Custom
Habit Names)](https://trello.com/c/cwQvDQI9/208-habits-v2-expected-habit-frequency-onlies-habit-maximum-custom-habit-names), @shalottastolat (shalott / astolat) has offered to work on part of it:
"Is anyone actively working on this (the expected frequency/counter/last time performed part)? I am finding that I want this a lot (I keep losing track myself) so would take a shot at it if no one else is on it!"

This issue is for discussing specifics.

### PR DESCRIPTION
**Pull request by shalott**
Fixes #8113 
### Changes

The original PR for this was here:
https://github.com/HabitRPG/habitica/pull/8175

Unfortunately while fixing a conflict in tasks.json, I messed up the rebase and wound up pulling in too many commits and making a giant mess, and I can't face the wrestling with git so am making a clean branch, sorry. :P

The only actual change here is to the test in cron.test.js, where it turns out that moment() and sinon.useFakeTimers() do not play well together. I should note that this means that the other tests that use the two together (eg, preening.test.js) are also not actually working as expected -- when using moment, it seems that sinon.useFakeTimers() is setting the time to the start of the UNIX epoch regardless of the day specified with moment. The fix for this is to use Date.new() instead. 

ORIGINAL BRANCH:

First pass at adding a counter to habits per discussion on the Trello board and in issue #8113. 

- For discussion: how to display the +/- counter. Currently, the counter on onlies is unmarked, and the counter on +/- habits is displayed as `+3|-2` I considered other options which are screenshotted in the issue description but this one looks like the best to me so far. 

- For discussion: I went with Monday as the start of the week per ISO standard and also because it seems more useful for separating weekend/weekday tasks. I could alternatively add a user pref for start of the week, however it occurs to me that this might be confusing to people who would expect it to be relevant in Dailies? 

Here's how the new feature works:

Habits have counterUp/counterDown and a frequency that is either daily, weekly, or monthly. The default is per day. The frequency can be set in the advanced options of habits. 

In tasks view, the user sees a counter on the habit box (to the left of the "top" icon, same place the streak counter goes on a daily) showing the number of times the habit has been ticked within the period.

If it is a +/- habit then it shows +#positive|-# negative per suggestion from @Alys. 

screenshot 2016-10-27 15 40 03

screenshot 2016-10-27 18 30 36

In the cron after the end of the counter period, the counter resets to 0.

---

UUID: 21a3965c-9203-4f83-b819-6b63902e03e2

### CODE REVIEW

File: test/api/v3/unit/libs/cron.test.js [hunk 1]
```diff
@@ -467,6 +467,67 @@ describe('cron', () => {
 
       expect(tasksByType.habits[0].value).to.equal(1);
     });
+
+    describe('counters', () => {
+      let notStartOfWeekOrMonth = new Date(2016, 9, 28).getTime(); // a Friday
```
Comment (shalott) [inline test/api/v3/unit/libs/cron.test.js:486]: Note: this is the line that makes the fake timer actually use the specified date.

**Comment by shalott** [comment]
Would appreciate suggestions for why the test is failing! Here's the error: 

`1) GET /export/history.csv should return a valid CSV file with tasks history data:
      AssertionError: expected 'daily 1,3621b015-bf4a-4b0a-a6a5-205fa52d3fb0,daily,2016-11-08 17:47:30,3.2' to equal 'habit 1,b7091e51-00f7-4063-9498-632071a90c12,habit,2016-11-08 17:47:29,1'
      + expected - actual
      -daily 1,3621b015-bf4a-4b0a-a6a5-205fa52d3fb0,daily,2016-11-08 17:47:30,3.2
      +habit 1,b7091e51-00f7-4063-9498-632071a90c12,habit,2016-11-08 17:47:29,1`

**Comment by shalott** [comment]
Update: the test was failing because the tasks created in the history test are sometimes not created in the exact order that they are passed to post. Another coder in the guild mentioned the same issue had occurred to them on a separate issue. In order to avoid future issues like this I have sorted the array of tasks post-creation.

**Comment by paglias** [comment]
@shalott thanks about your latest commit! I may merge it immediately before the rest of the PR because it's very useful

**Comment by shalott** [comment]
Great, glad it helps!

**Comment by Alys** [comment]
@shalott I've given you one tier already for your improvement to the history tests in your merged commit https://github.com/HabitRPG/habitica/commit/8c4f35daf4e7fcea71fcbfe9e076588bb664db8d, since it's clearly useful enough to stand alone. :)

The remaining code in this PR might be suitable for a second tier if it is merged, since the first couple of Blacksmith tiers are each often awarded for relatively small pieces of work. 

You can read more about [contributor tiers and rewards here](http://habitrpg.wikia.com/wiki/Contributor_Rewards). Note that as your tier increases, it takes progressively more work to achieve the next one, so for the higher tiers, larger PRs or a few PRs are required. There's some information about that at Contributor Rewards, but feel free to ask here if you have questions!

**Comment by shalott** [comment]
Hey, I looked around and I'm still seeing the sporadic test error popping up in other pullreqs despite the fix, so apparently it's not working? My best guess now is that in these failures the tasks aren't all actually getting created/returned (which error isn't visible because the test fails on a mismatch before it gets to the later elements of the array). Anyway, I'm happy to take another shot at it in a separate pullreq, but I'm traveling through the Thanksgiving holiday so it won't be until the following week.

**Comment by crookedneighbor** [comment]
Whenever a PR is opened, it runs 2 builds. One against the branch itself and one against the branch as if it had been merged into develop. It's possible that the builds that are still failing just haven't pulled in the latest changes with your fix.

**Comment by paglias** [comment]
It's just happened in `develop` so the bug still exists :(

**Comment by paglias** [comment]
Now that I'm looking at the code of your commit again, it sorts `tasks` but that's a local variable in the test that is not used anymore. It shouldn't affect the database order

**Comment by shalott** [comment]
No the tasks array is used in order to get the task ids to score them and also to do the check. I think it has to be that the post isn't actually returningthe expected array of tasks sometimes. I would start by doing expect statements to check that the tasks array returned by post does have all the expected tasks.

**Comment by paglias** [comment]
@shalott I'm commenting out the test because a majority of builds are failing now, but if you want to propose a new PR it would be great!

**Comment by TheHollidayInn** [comment]
Looks like the builds are good now. What is the current status of this PR?

**Comment by shalott** [comment]
Have resolved the conflicts just now, lmk what if any other changes you need.

**Comment by TheHollidayInn** [comment]
Alright, I have pulled this locally and tested. We are discussing a release plan, but this should be merged in soon!

**Comment by TheHollidayInn** [comment]
Noted towards your Tier! Thank you!

**Comment by Alys** [comment]
Just FYI for the record:

We're having some trouble on our test server with the PRs that have been merged in recently, so I've started testing each one individually.  I tested this one on my local install by fetching upstream, then checking out 6d0df7844168916c2a948e521eff84e73b2be9fa and doing `npm start`. My testing has been brief but everything looks good here so far. My tests show:

- Habits can be scored up and down (with the new counters incrementing correctly - I think I'm really going to like them!)
- Tasks can be edited
- Tasks can be deleted
- Tasks can be added, and if the "edit mode" setting is turned on, the edit modal appears as soon as you add the task
- The day/week/month setting for clearing the counters can be changed and the change sticks after editing.
- Challenges can be created (only very brief testing done).
- A Habit in a challenge can have the day/week/month counter reset field edited by the challenge owner. The setting is saved correctly. 
- A challenge participant can NOT edit the counter reset field on challenge tasks, which is good.

I have NOT tested the following things but they can be done once our test server is behaving again:
- Whether the counter reset performs correctly during cron (although I have seen nothing to indicate it doesn't)
- How this interacts with group plans.

We should probably do a bit more thorough testing of challenges, just to be sure.

**Comment by Alys** [comment]
Here's some extra details about how this works, for the benefit of Socialites and wiki editors. Some of this might already be described above.

- Habits display a counter
- Habits can display both positive and negative counter
- The mobile apps will NOT display the counter (until the apps are updated) but all clicks WILL be recorded in the counter and will become visible in the website (this applies to clicks made in the apps, the Android widget, and in any third-party tools that increment Habits).
- Counter resets based on “Counter Reset Frequency” in Advanced Options in the Habit’s edit screen. 
   - Daily counters reset each time cron runs.
   - Weekly counters reset when you cron on Monday morning (or on the next day you use Habitica if you don’t use Habitica on Monday). 
   - Monthly counters reset when you cron on the first of the month (or on the next day you use Habitica if you don’t use Habitica on the first).
- All existing Habits have been given the “Daily” value for the Counter Reset Frequency (you can edit them to change that).
- The Counter Reset Frequency can NOT be changed for Habits from active challenge or from group plans.
   - Challenge owners and group plan owners CAN set the reset frequency.

**Comment by Alys** [comment]
I've increased @shalott's tier to 2 for this. Tier 1 was given as I described in https://github.com/HabitRPG/habitica/pull/8198#issuecomment-261758717 for improvement to the history tests. The rest of this PR (implementing the Habits counter) is certainly worth a tier by itself.

**Pull request merged**